<a href="https://colab.research.google.com/github/ArthisreeRamesh/GenAI-Agentic-AI/blob/main/Autogen.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
pip install autogen

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 881.2/881.2 kB 50.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 3.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 147.8/147.8 kB 12.2 MB/s eta 0:00:00


In [ ]:
from getpass import getpass
Open_AI_Key = getpass("enter key")

enter key··········


In [ ]:
config_list_gpt = {
    "config_list": [{"model":"gpt-4o-mini","api_key":Open_AI_Key}],
    "temperature": 0
}

human_input_mode = never,terminate,always ,name must be related to task preferably so system will understand

In [ ]:
import autogen
from autogen import ConversableAgent
from IPython.display import display, Markdown

/usr/local/lib/python3.12/dist-packages/google/cloud/aiplatform/models.py:52: FutureWarning: Support for google-cloud-storage < 3.0.0 will be removed in a future version of google-cloud-aiplatform. Please upgrade to google-cloud-storage >= 3.0.0.
  from google.cloud.aiplatform.utils import gcs_utils


In [ ]:
agent_mrk_str = ConversableAgent(
    name="marketing_strategist",
    system_message="""You are an expert marketing strategist for both B2B and B2C side of a business.
                      Your expertise lies in creating strategies for companies and products in the 'Education Tech industry'.""",
    llm_config=config_list_gpt,
    human_input_mode="NEVER"
)

In [ ]:
# Define the task for the agent(like giving input to llm)
content = """I want some solid marketing strategies for B2B promotions for my 'Project Management Courses'."""
reply = agent_mrk_str.generate_reply(messages=[{"content":content, "role":"user"}])
display(Markdown(reply))

Two Way Agent - Agentic AI with AutoGen
Problem Statement:
You are the CMO - Chief marketing officer of an EduTech company specialized in project management courses.

CEO - Chief Execuative Officer wants to focus on B2C while you (CMO) wants to focus on B2B

Simulate a conversation between the CEO and CMO


In [ ]:
CMO_agent = ConversableAgent(
    name="CMO",
    system_message="""You are a Chief Marketing Officer in an Edu Tech Company.
                      You are discussing with the CEO whether you should focus on B2B or B2C for project management courses.
                      As the CMO, you want to focus on B2B.""",
    llm_config=config_list_gpt,
    human_input_mode="NEVER"
)

In [ ]:
CEO_agent = ConversableAgent(
    name="CEO",
    system_message="""You are a Chief Executive Officer in an Edu Tech Company.
                      You are discussing with the CMO whether you should focus on B2B or B2C for project management courses.
                      As the CMO, you want to focus on B2B.""",
    llm_config=config_list_gpt,
    human_input_mode="NEVER"
)

In [ ]:
messages = ["I think we should focus more on B2B besiness for our project management courses."]

silent=False->	You see every message exchanged between the two agents printed in your notebook/console.	Like watching the meeting live
silent=True->	They talk in the background (no messages shown), and you only get the final result.	Like they had a private meeting and gave you only the summary

"reflection_with_llm" means:
The agents themselves reflect on the conversation using the LLM and then generate a smart summary.
Instead of just joining all the messages or taking the last line,
the model thinks about:

What was discussed,decisions made ,insights shared

In [ ]:
result = CMO_agent.initiate_chat(CEO_agent, message=messages[0],
                                 max_turns=5, silent=False, summary_method="reflection_with_llm")

CMO (to CEO):

I think we should focus more on B2B besiness for our project management courses.

--------------------------------------------------------------------------------
CEO (to CMO):

As the CEO, I appreciate your perspective on focusing on B2B for our project management courses. Let's discuss the potential benefits and challenges of this approach.

1. **Market Demand**: The demand for project management training in businesses is growing, especially as companies look to improve efficiency and productivity. We can tap into this market by offering tailored solutions that meet specific organizational needs.

2. **Long-term Contracts**: B2B relationships often lead to long-term contracts, which can provide us with a more stable revenue stream compared to B2C, where customer acquisition can be more volatile.

3. **Scalability**: By focusing on B2B, we can scale our offerings more effectively. Once we establish a relationship with a company, we can offer additional courses or servic

In [ ]:
len(result.chat_history)

10

Problem Statement - Sequential Conversable Chat
Role: Strategist at a leading e-commerce company
Task: Prepare for the upcoming festive season by analyzing completitor's discounts and promptional offers.
Objective: Design and compare a competitive strategy to either match or surpass the competitors offer
Solution : Develope Agentic AI System with the following approach :
a Dynamic coding

b Web Scraping

c Data Summarization

Outcome: Enable data-driven decision-making to ensure a competitive edge during the festive season


In [ ]:
import autogen
from autogen import ConversableAgent, AssistantAgent, UserProxyAgent
from autogen.coding import LocalCommandLineCodeExecutor
from IPython.display import display, Markdown, Image, HTML

/usr/local/lib/python3.12/dist-packages/google/cloud/aiplatform/models.py:52: FutureWarning: Support for google-cloud-storage < 3.0.0 will be removed in a future version of google-cloud-aiplatform. Please upgrade to google-cloud-storage >= 3.0.0.
  from google.cloud.aiplatform.utils import gcs_utils


UserProxyAgent is like a substitute for user to initiate conversation/assign task

In [ ]:
# Initiate the conversation and assigning task to the agent
user_proxy = UserProxyAgent(
    name="User",
    is_termination_msg=lambda msg: msg.get("content") is not None and "TERMINATE" in msg['content'],
    human_input_mode="TERMINATE",
    code_execution_config=False,
    llm_config=False
)

In [ ]:
# This agent will write a code for web scrapping
Code_writer_Agent = ConversableAgent(
    name="Code_writer",
    system_message="""You are a Python Developer. You use your coding skill to solve problems.
                      Once the task is done, returns 'TERMINATE'.""",
    llm_config=config_list_gpt,


)

In [ ]:
# This agent will evaluate the codewrite code and provide their suggestion
Code_reviewer_Agent = ConversableAgent(
    name="Code_reviewer",
    system_message="""You are an experienced python developer. You review the code given to you based on the requirements.
                      Once the task is done, returns 'TERMINATE'.""",
    llm_config=config_list_gpt,


)

In [ ]:
import os
os.getcwd()

'/content'

In [ ]:
local_executor = LocalCommandLineCodeExecutor(timeout=20, work_dir='./code_files')

In [ ]:
# This agent will save code for both writer and reviewer
local_executor_agent = ConversableAgent(
    name="LocalExecutorAgent",
    llm_config=False,
    code_execution_config={"executor":local_executor},
    human_input_mode="ALWAYS")

In [ ]:
messages = ["""To check whether there are any offers or discounts available on a given
              e-commerce website -
              https://www.flipkart.com/

              Follow these steps,
              1. Download the html page on the given above url
              2. We only need text contents, so remove any Image content, JavaScript, CSS and other details
              3. Save the remaining html content.
              """,

            """The code is given by an LLM. Review the code based on the above requirements and provide it for execution""" ]

In [ ]:
chat_result = autogen.initiate_chats(
    [{
        "sender": user_proxy,
        "recipient": Code_writer_Agent,
        "message": messages[0],
        "summary_method": "last_msg",
    },
    {
        "sender": local_executor_agent,
        "recipient": Code_reviewer_Agent,
        "message": messages[0] + messages[1],
        "summary_method": "last_msg",
    }]
)


********************************************************************************
Starting a new chat....

********************************************************************************
User (to Code_writer):

To check whether there are any offers or discounts available on a given
              e-commerce website -
              https://www.flipkart.com/

              Follow these steps,
              1. Download the html page on the given above url
              2. We only need text contents, so remove any Image content, JavaScript, CSS and other details
              3. Save the remaining html content.
              

--------------------------------------------------------------------------------

>>>>>>>> USING AUTO REPLY...
Code_writer (to User):

To accomplish the task of checking for offers or discounts on the Flipkart website, we can use Python with libraries such as `requests` for downloading the HTML content and `BeautifulSoup` from `bs4` for parsing and cleaning the HTML.

In [ ]:
import requests
from bs4 import BeautifulSoup
import random
import time

# List of User-Agent strings
user_agents = [
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36",
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.1.1 Safari/605.1.15",
    "Mozilla/5.0 (Linux; Android 10; SM-G973F) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.77 Mobile Safari/537.36",
    # Add more User-Agent strings as needed
]

# Step 1: Download the HTML page
url = "https://www.flipkart.com/"
headers = {
    "User-Agent": random.choice(user_agents)  # Randomly select a User-Agent
}

response = requests.get(url, headers=headers)

# Check if the request was successful
if response.status_code == 200:
    html_content = response.text

    # Step 2: Parse the HTML and remove unwanted content
    soup = BeautifulSoup(html_content, 'html.parser')

    # Remove scripts and styles
    for script in soup(['script', 'style']):
        script.decompose()

    # Get text content
    text_content = soup.get_text()

    # Step 3: Save the remaining HTML content
    with open('flipkart_offers.txt', 'w', encoding='utf-8') as file:
        file.write(text_content.strip())
else:
    print("Failed to retrieve the webpage.")

# Introduce a delay to mimic human behavior
time.sleep(random.uniform(1, 3))  # Sleep for 1 to 3 seconds

# Indicate that the task is complete
print("")

Failed to retrieve the webpage.



In [ ]:
import requests
from bs4 import BeautifulSoup
import time

# Step 1: Download the HTML page
url = "https://www.flipkart.com/"
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
}

try:
    response = requests.get(url, headers=headers)

    # Check if the request was successful
    if response.status_code == 200:
        html_content = response.text

        # Step 2: Parse the HTML and remove unwanted content
        soup = BeautifulSoup(html_content, 'html.parser')

        # Remove scripts and styles
        for script in soup(['script', 'style']):
            script.decompose()

        # Get text content
        text_content = soup.get_text(separator='\n')

        # Step 3: Save the remaining HTML content
        with open('flipkart_offers.txt', 'w', encoding='utf-8') as file:
            file.write(text_content)

        print("Content saved successfully.")
    else:
        print(f"Failed to retrieve the webpage. Status code: {response.status_code}")

except requests.exceptions.RequestException as e:
    print(f"An error occurred: {e}")

# Optional: Introduce a delay to avoid overwhelming the server
time.sleep(2)

# Indicate that the task is complete
print("TERMINATE")

Failed to retrieve the webpage. Status code: 529
TERMINATE


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Failed to retrieve the webpage. Status code: 529
TERMINATE
site overloaded

In [ ]:
import requests
from bs4 import BeautifulSoup

# Define headers
headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.3"
}

# Make a GET request to the website with the defined headers
response = requests.get('https://www.flipkart.com/', headers=headers)
html_content = response.text

soup = BeautifulSoup(html_content, 'lxml')

# Removing unnecessary content: script and style tags
for script in soup(["script", "style"]):
    script.extract()

# Get text content
text_content = soup.get_text(strip=True)

# Saving the remaining content
with open("flipkart.txt", "w", encoding="utf-8") as file:
    file.write(text_content)

In [ ]:
# Let's parse the uploaded flipkart.txt file to extract product categories and their discount percentages.
# We'll look for patterns like "Upto 80% Off" or "Up to 70% Off" and create a table.

import re
import pandas as pd

# Read the file content
file_path = "flipkart.txt"
with open(file_path, "r", encoding="utf-8") as file:
    text = file.read()

# Regex to match items with discounts like "Upto 80% Off" or "Up to 70% Off"
pattern = re.compile(r"([A-Za-z0-9 &]+?)\s*(?:Upto|Up to)\s*(\d+)% Off", re.IGNORECASE)

matches = pattern.findall(text)

# Create a DataFrame
df = pd.DataFrame(matches, columns=["Product/Category", "Discount (%)"])

# Remove duplicates and clean up whitespace
df["Product/Category"] = df["Product/Category"].str.strip()
df = df.drop_duplicates().reset_index(drop=True)

import caas_jupyter_tools
caas_jupyter_tools.display_dataframe_to_user(name="Flipkart Discounts Table", dataframe=df)


Group Chat
Problem statement
New Project : A mobile gaming company is developing a game for PS7, requiring new expertise

HR Role: The HR team is responsible for determingin hiring requirements by collaborating with relevant departments.

Agentic System : Implement a group chat system where HR initiates discussion with a Group Chat Manager to gather hiring needs from team member

In [ ]:
from autogen import GroupChat, GroupChatManager

In [ ]:
hr_agent = ConversableAgent(
    name="hr_agent",
    system_message = """You are the head of HR in a company. The company is initiating the development of
                        a new B2C tech product and aims to assemble a dedicated team for this project.
                        Gather the hiring requirements from different departments as needed.""",
    llm_config = config_list_gpt,
    human_input_mode="NEVER")

In [ ]:
product_agent = ConversableAgent(
    name="product_agent",
    system_message = """You are the head of Product in a company developing a new B2C tech product.
                Your goal is to communicate with the HR department to outline the hiring requirements for
                assembling a dedicated Product Management team for this project.
                Clearly and concisely list the key roles, qualifications,
                and skills and budget needed for this initiative.""",
    llm_config = config_list_gpt,
    human_input_mode="NEVER")

In [ ]:
engineering_agent = ConversableAgent(
    name="engineering_agent",
    system_message = """You are the head of Engineering in a company developing a new B2C tech product.
                Your goal is to communicate with the HR department to outline the hiring requirements for
                assembling a dedicated Engineering team for this project.
                Clearly and concisely list the key roles, qualifications,
                and skills and budget needed for this initiative.""",
    llm_config = config_list_gpt,
    human_input_mode="NEVER")

In [ ]:
marketing_agent = ConversableAgent(
    name="marketing_agent",
    system_message = """You are the head of Marketing in a company developing a new B2C tech product.
                Your goal is to communicate with the HR department to outline the hiring requirements for
                assembling a dedicated Marketing team for this project.
                Clearly and concisely list the key roles, qualifications,
                and skills and budget needed for this initiative.""",
    llm_config = config_list_gpt,
    human_input_mode="NEVER")

In [ ]:
operation_agent = ConversableAgent(
    name="operation_agent",
    system_message = """You are the head of Operation in a company developing a new B2C tech product.
                Your goal is to communicate with the HR department to outline the hiring requirements for
                assembling a dedicated Operation team for this project.
                Clearly and concisely list the key roles, qualifications,
                and skills and budget needed for this initiative.""",
    llm_config = config_list_gpt,
    human_input_mode="NEVER")

In [ ]:
customer_support_agent = ConversableAgent(
    name="customer_support_agent",
    system_message = """You are the head of Customer Support in a company developing a new B2C tech product.
                Your goal is to communicate with the HR department to outline the hiring requirements for
                assembling a dedicated Customer Support team for this project.
                Clearly and concisely list the key roles, qualifications,
                and skills and budget needed for this initiative.""",
    llm_config = config_list_gpt,
    human_input_mode="NEVER")

In [ ]:
group_chat = GroupChat(
    agents=[customer_support_agent,operation_agent,product_agent,engineering_agent,marketing_agent],
    messages = [],
    max_round = 7,
    speaker_selection_method = 'auto',
    select_speaker_auto_llm_config= config_list_gpt
)
#every agent here needs llm, speaker_selection_method =auto because we are not specifying in which order all should speak
#we just mention hr will initiate in last step

In [ ]:
group_chat_manager = GroupChatManager(groupchat = group_chat,)

In [ ]:
chat_result = hr_agent.initiate_chat(
    group_chat_manager,
    message = """A gaming company focussed on mobile games which is launching a new game designed for PS7.
                Please provide hiring requirement for this new product.""",
    summary_method ="last_msg",)
#hr initiates the conversation
#summary_method =last msg  -we will only get the conclusion of thier conversations not every conversation

hr_agent (to chat_manager):

A gaming company focussed on mobile games which is launching a new game designed for PS7. 
                Please provide hiring requirement for this new product.

--------------------------------------------------------------------------------

Next speaker: product_agent

product_agent (to chat_manager):

Subject: Hiring Requirements for Product Management Team for New PS7 Game

Dear HR Team,

As we embark on the development of our new mobile game designed for the PS7, it is crucial to assemble a dedicated Product Management team to ensure the project's success. Below are the key roles, qualifications, skills, and budget considerations for this initiative.

### Key Roles and Responsibilities:

1. **Product Manager (2 Positions)**
   - **Qualifications**: 
     - Bachelor’s degree in Business, Marketing, or a related field; MBA preferred.
     - 3-5 years of experience in product management, preferably in gaming or tech.
   - **Skills**:
     - Strong unde